First draft of missing data guide (#12)
egap/methods-guides: missing_data/missing_data.Rmd

---
title: "10 Things to Know about Missing Data"
author: "Tara Slough"
date: "2/28/2019"
output: html_document
---
##1. What is missing data?

When variables are missing some data values, we say that there is "missing data." Depending on your software and the coding of the dataset, missing values may be coded as `NA`, `.`, an empty cell (`""`), or a common numeric code (often `-99` or `99`).

> It is useful to distinguish missing data by variable type since the consequences of missing data for estimation and interpretation vary. For our purposes, we will consider three types of variables: pretreatment covariates, treatment indicator(s), and outcome (or dependent) variables. Pretreatment covariates are variables that we observe and measure before treatment is assigned. Outcome (or dependent) variables refer to outcomes that are measured after the assignment of treatment. 

Reviewer: ```suggestion
The consequences of missing data for estimation and interpretation depend on the type of variable missing the data. For our purposes, we will consider three types of variables: pretreatment covariates, treatment indicator(s), and outcome (or dependent) variables. Pretreatment covariates, often known simply as "covariates", are variables that we observe and measure before treatment is assigned. Outcome (or dependent) variables refer to outcomes that are measured after the assignment of treatment. 
```

> Missing data emerges for different reasons. In survey data, a respondent could decline to answer a question or quit the survey without completing all questions. Moreover, in a panel survey, some subjects may not be relocated for the second or later wave. With administrative data, records may be lost at some point in the process of collecting or recording data. To the extent that we can know the process by which data becomes missing, we can better understand the consequences of missing data for our analysis and inferences.

Reviewer: ```suggestion
Missing data emerges for different reasons. In survey data, a respondent could decline to answer a question or quit the survey without completing all questions. In a panel survey, some subjects may skip the second or later waves. With administrative data, records may be lost at some point in the process of collecting or recording data. To the extent that we can know the process by which data becomes missing, we can better understand the consequences of missing data for our analysis and inferences.
```

> Missing data emerges for different reasons. In survey data, a respondent could decline to answer a question or quit the survey without completing all questions. Moreover, in a panel survey, some subjects may not be relocated for the second or later wave. With administrative data, records may be lost at some point in the process of collecting or recording data. To the extent that we can know the process by which data becomes missing, we can better understand the consequences of missing data for our analysis and inferences.

Reviewer: @taraslough Just seeing a fair amount of passive voice in this early draft. I've been trying linters to help with writing style and wonder if that might be useful here too: https://hackernoon.com/lint-lint-and-away-linters-for-the-english-language-70f4b22cc73c, https://atom.io/packages/linter-write-good, https://ben.balter.com/2016/12/23/atom-for-prose/.  I edit using vim so I install the ale linter (https://github.com/w0rp/ale). You can also cut and paste into Hemmingway (http://www.hemingwayapp.com/). I'll try to stop the style editing now and just read for content.

Reviewer: I also just found https://github.com/ropenscilabs/gramr for folks editing in RStudio.

##2. Missing treatment or outcome data can bias our ability to describe empirical patterns and estimate causal effects.

> One perverse consequence of missing data is that it can induce bias in our estimates of descriptive patterns and causal effects. Consider a researcher trying to describe the income distribution in a country. Some individuals' incomes are missing but the researcher describes the nonmissing data at hand. Suppose low-income individuals are less likely to report their income than high-income individuals, thus missingness is concentrated in the lower portion of the distribution. Then, the researcher's characterization of the income distribution is apt to be biased. For example, the researcher's estimate of median income is apt to be higher than the true (unknown) median income because more data is missing from the lower portion of the distribution. Because missingness is correlated with the variable we are trying to describe, our characterization of the median of the distribution is biased. 

Reviewer: ```suggestion
One perverse consequence of missing data is that it can induce bias in our estimates of descriptive patterns and causal effects. Consider a researcher trying to describe the income distribution in a country. Some individuals' incomes are missing but the researcher describes the nonmissing data at hand. Suppose low-income individuals are less likely to report their income than high-income individuals, thus missingness is concentrated in the lower portion of the income distribution. Then, the researcher's characterization of the income distribution is apt to be biased. For example, the researcher's estimate of median income is apt to be higher than the true (unknown) median income because more data is missing from the lower portion of the distribution. Because missingness is correlated with the variable we are trying to describe, our characterization of the median of the distribution is biased. 
```

Reviewer: Maybe say something about bias in description to not set people up to expect that the example should relate to causal effects?

We can illustrate two types of missingness by considering a variable $x$ -- perhaps income, as above. The 

> left panel shows the full distribution of the variable $x$ without missing data. The middle panel depicts a scenario in which simulated missingenss in $x$ is not independent of $x$ -- low values of $x$ are much more likely to be missing. The right panel depicts a scenario in which simulated missingness in $x$ is independent of $x$ (often called "missing at random"). The red line represents the median of the distribution without missingness. The blue lines represent the median of the observed data in each simulation. Where missingness is not independent of $x$, we observe that the median of the nonmissing data is, in this instance, higher than the true median. This illustrates that missing data can bias our characterization of..

Reviewer: ```suggestion
left panel shows the full distribution of the variable $x$ without missing data. The middle panel depicts a scenario in which simulated missingenss in $x$ is not independent of $x$ -- low values of $x$ are much more likely to be missing. The right panel depicts a scenario in which simulated missingness in $x$ is independent of $x$ (often called "missing at random"). The red line represents the median of the distribution without missingness. The blue lines represent the median of the observed data in each simulation. Where missingness is not independent of $x$, we observe that the median of the nonmissing data is, in this instance, higher than the true median. This illustrates that missing data can bias our descriptions of single variables.
```

> ```{r, warning = F, message = F, echo = F, fig.caption = "Two types of missingness and implications for estimation of the median. The red line "}

Reviewer: @malisi  I like the idea of having the "code" button available to reveal or hide the code. I suspect that people will be happy to learn how to made this plot.

simulation <- function(){
  Y0 <- rnorm(2000)
  Z <- complete_ra(2000)
  Yobs <- Y0 + Z * .5 # .5 standard deviation treatment effect
  Z_miss <- rbinom(n = 2000, prob = Z * pnorm(Yobs), size = 1)
  Z_miss_ind <- rbinom(n = 2000, prob = sum(Z_miss)/2000, size = 1)
  Y_miss <- rbinom(n = 2000, prob = Z * pnorm(Yobs), size = 1)
  Y_miss_ind <- rbinom(n = 2000, prob = sum(Y_miss)/2000, size = 1)
  
  m1 <- lm_robust(Yobs ~ Z, subset = Z_miss == 0)
  m2 <- lm_robust(Yobs ~ Z, subset = Z_miss_ind == 0)
  m3 <- lm_robust(Yobs ~ Z, subset = Y_miss == 0)
  m4 <- lm_robust(Yobs ~ Z, subset = Y_miss_ind == 0)

  return(sapply(list(m1, m2, m3, m4), FUN = extract_ests))
}

reps <- replicate(n = 500, expr = simulation())

data.frame(ests = as.vector(t(reps)),
           missing = rep(c("Missing Treatment Indicator", "Missing Outcome"), each = 1000),
           pos = rep(c("Missing is\nNot Independent of POs", "Missingness is\nIndependent of POs"), each = 500)) %>%
  ggplot(aes(x = ests)) + 
  facet_grid(pos ~ missing) +
  geom_histogram(bins = 100) +
  geom_vline(xintercept = .5, col = "red") +
  theme_minimal() +
  xlab("ATE Estimates")

> ```

Reviewer: Great graph and discussion!

|-----------------------|:-------:|:--------:|:--------:|
| Always Reporter       | $\pi_A$ |     1    |     1    |
| If Treated Reporter   | $\pi_T$ |     1    |     0    |
| If Untreated Reporter | $\pi_U$ |     0    |     1    |
| Never Reporter        | $\pi_N$ |     0    |     0    |

In the case in which missingness is related to potential outcomes but not to treatment assignment, we are generally not able to identify an unbiased estimate of the ATE. However we may be interested in identifying the LATE among always reporters -- those subjects for which we observe the outcome regardless of treatment assignment. This effect can be a main estimand of interest and even the most policy-relevant estimand in certain settings. However, to estimate a LATE, we want to be sure that the outcomes that we observe are those of always reporters, not if-treated or if-untreated reporters. 

To assess the plausibility of ths conjecture, we often examine the relationship between treatment assignment and reporting (non-missingness). Using the notation from the table, we can estimate:

\begin{align}
E[R_i(1)-R_i(0)] &= \pi_A + \pi_T - (\pi_A + \pi_U)\\
&=\pi_T - \pi_U
\end{align}

> If we find no difference in reporting across treatment groups, this test provides no evidence that $\pi_T \neq \pi_U$. However, to further justify that the complete observations in the data are those of "always reporters," we further must know that $\pi_T = \pi_U = 0$. We cannot test this by examining the relationship between reporting and treatment assignment. As such, we generally complement this test with a justification of why reporting should not be endogenous to treatment assignment. If reporting is not endogenous, then in principle the assumption that $\pi_T = \pi_U = 0$ holds. As such, if we find no evidence of differential levels of reporting and see no way that reporting should be endogenous to treatment, we can justify that the $E[Y_i(1)|R_i(1) = 1] - E[Y_i(0)|R_i(0) = 1]$ estimates the LATE among always reporters.

Reviewer: ```suggestion
If we find no difference in reporting across treatment groups, this test provides no evidence that $\pi_T \neq \pi_U$. However, to further justify that the complete observations in the data are those of "always reporters," we further must know that $\pi_T = \pi_U = 0$. We cannot test this by examining the relationship between reporting and treatment assignment. As such, we generally complement this test with a justification of why reporting should not be endogenous to treatment assignment: for example by assessing covariate balance between treated and control units who are not missing, or by assessing covariate balance between units with missing vs non-missing outcomes. If reporting is not endogenous, then in principle the assumption that $\pi_T = \pi_U = 0$ holds. As such, if we find no evidence of differential levels of reporting and see no way that reporting should be endogenous to treatment, we can justify that the $E[Y_i(1)|R_i(1) = 1] - E[Y_i(0)|R_i(0) = 1]$ estimates the LATE among always reporters.
```

> [Jake, I think you wanted this as a justification of MIPO too (when adding covariates) and doing RI on f-statistics or the like. Do you still want this?]

Reviewer: ```suggestion

```

Reviewer: I added a phrase to just gesture at the common practice in vague terms rather than arguing for the d^2 test.

##7. What is imputation?

The most common approaches used to deal with missing data involve the imputation or "filling in" of missing values. In the following dataset with missingness, imputation procedures "fill in" the missing data -- the `NA`s. 

```{r, warning = F, message = F, echo = F}
set.seed(1234)

> X <- rnorm(12)

Reviewer: ```suggestion
X <- signif(rnorm(12),4)
```

Reviewer: Just trying to make the table a little easier to read --- no need for so many digits. Maybe better to do this in `kable()`.

##8. How do we address missingness of pre-treatment covariates and why does this matter?

As mentioned in #4, we should never "drop" observations on account of missing pre-treatment data. In order to estimate a model with covariate adjustment, thus, we need to "fill in" missing values to avoid dropping observations. We outline two forms of imputation advocated for missing pre-treatment covariates. The most common approach to address missingness of pre-treatment covariates is to create indicators for missingness and include these as covariates. To do this form of imputation:

1. Substitute a numerical value for the `NA` (as necessary). In the dataset below, we impute a `0` for all values of `Xobs` that are `NA`s. 
2. Create an indicator for the missingness in each pretreatment variable. This indicator, `X_missing_ind` takes the value `1` whenever `Xobs` is NA, and a 1 otherwise.

> ```{r, warning = F, message = F, echo = F}

Reviewer: Some funkiness in the headers of this table in the html.

Reviewer: Also too many digits.

2. A **minimum** bound by imputing the *minimum* value of the outcome for missing values *in treatment* ($Z=1$) and imputing the *maximum* value of the outcome for missing values *in control* ($Z=0$).

Using the above dataset, we can construct two variables `Y_maxbound` and `Y_minbound` as follows:
```{r, warning = F, message = F, echo = F}
Y_maxbound <- ifelse(is.na(Yobs) & Z == 1, 4,
                     ifelse(is.na(Yobs) & Z == 0, 1, Yobs))
Y_minbound <- ifelse(is.na(Yobs) & Z == 1, 1,
                     ifelse(is.na(Yobs) & Z == 0, 4, Yobs))

knitr::kable(data.frame(Xobs, Z, Yobs, Y_maxbound, Y_minbound))
```

Without covariate adjustment, we can obtain our interval estimate of the ATE by estimating:

```{r, warning = F, message = F}
upper <- lm(Y_maxbound ~ Z)
lower <- lm(Y_minbound ~ Z)

coef(upper)[2]
coef(lower)[2]
```

Our interval estimate of the ATE using Manski bounds is thus [`r round(coef(lower)[2], 2)`, `r round(coef(upper)[2], 2)`].

##10. Multiple imputation for missing outcomes allows for point estimation of ATEs, but relies on stronger assumptions than bounding.

> The methods in #8 and #9 describe methods of single imputation, where a single value is substituted for missing values. In multiple imputation, we impute missing values of the dataset multiple times according to an assumed stochastic data generating process. Different methods for multiple imputation impose different structures and assumptions data generating processes used to impute missing values. In general, multiple imputation proceeds via three stages:

Reviewer: s
```suggestion
The methods in #8 and #9 describe methods of single imputation, where a single value is substituted for missing values. In multiple imputation, we impute missing values of the dataset multiple times according to an assumed stochastic data generating process. Different methods for multiple imputation impose different structures and assumptions about the probability distributions governing the data generating processes used to impute missing values.  In general, multiple imputation proceeds via three stages:
```
